DRILL-6880: For Hash-Join hash-table build - treat null keys as an equal match

Ben-Zvi opened the pull request:
When a new key is inserted into the hash-table's, a matching bucket is found (i.e., from its hash value modulo the table's size), but that bucket may contain an old key(s). A generated code method ("isKeyMatchInternalBuild") is used to compare and find if the two keys are identical.

Previously a special check in that method addressed the case where (any pair of matching) columns where both null; returning "no match" if so.  In such a case, the hash-join would have entered the new key as a new entry at that bucket, leading to a growing useless chain that could end up in O(n^2) work (as each new key checks against all old ones).

  *The fix*: Generate a new method - "areBothKeysNull" - to perform that special check prior to the actual key match; used only for hash-join during build time.  Then this method returns the **inverse logic** - return a **"match"** when any two matching key-columns are both null.  This would avoid the useless chain in the hash-table. (A linked list would be created in the Hash-Join Helper, but that's an O(1) work that would be too complex/inefficient to avoid).

   Also note that the case of the keys like: <1,null>, <2,null>, <3,null>, <4,null>, ...... would not be addressed by the above change, as most of these keys would end up in different buckets, but that is still O(1) work, and no long chain.

  *The implementation*: Customize the internal method _setupIsKeyMatchInternal_ to also generate the new method _areBothKeysNull_ . The other methods it generates should not change from before, except _isKeyMatchInternalBuild_ where the special "both null" check is now eliminated.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/common/ChainedHashTable.java:
```diff
@@ -208,7 +216,10 @@ public HashTable createAndSetupHashTable(TypedFieldId[] outKeyFieldIds) throws C
     }
 
-
+    // Only in case of an equi-join: Generate a special method to check if both the new key and the existing key (in this HT bucket) are nulls
```
When you say equi-join, do you mean inner join ?  The join operators don't handle non-equijoins currently.

Ben-Zvi replied:
The 'equi-join' term was just following the 'Comparator.EQUALS'; and is correct as we only support these now.
   The point here is that for *all joins*, nulls are equal on build/insert, but not equal on probe. This works the same for outer joins, as an unmatched probe on a single hash-table key is the same as un-matching multiple such keys.  Maybe I should add this to the comment to clarify.

Ben-Zvi commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/common/ChainedHashTable.java:
```diff
@@ -235,9 +246,14 @@ public HashTable createAndSetupHashTable(TypedFieldId[] outKeyFieldIds) throws C
   private void setupIsKeyMatchInternal(ClassGenerator<HashTable> cg, MappingSet incomingMapping, MappingSet htableMapping,
       LogicalExpression[] keyExprs, List<Comparator> comparators, TypedFieldId[] htKeyFieldIds) {
+
+    boolean isProbe = incomingMapping == KeyMatchIncomingProbeMapping;
```
The initial implementation did use flags, but looked a little ugly with "true" and "false" parameters in each call. I'll try and create an enum to make it cleaner.

Ben-Zvi commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/common/HashTableTemplate.java:
```diff
@@ -224,6 +224,13 @@ private boolean isKeyMatch(int incomingRowIdx,
+        // be used (not putting it into the helper would take a bigger code change, and some performance cost, hence
+        // not worth it).  In the past such a new null key was added into the Hash-Table (i.e., no match), which
+        // created long costly chains - SEE DRILL-6880)
+        if ( areBothKeysNull(incomingRowIdx, currentIdxWithinBatch) ) { return true; }
```
The code on old line 230 is in a way "not clean", as it violates the method's name/signature (which implies just a functional comparison of two keys). That code advances to the next key in the hash-table chain in case of a no-match. The "clean way" would be to make this code a separate "advanceToNextIdx" method, which actually fits well in the for() loop where _isKeyMatch()_ is used.

amansinha100 replied:
Agree that the function name does not reflect the fact that the index gets incremented.  Either we change the function name to `isKeyMatchAndUpdateIndex()` or do that adjustment outside this function.

Ben-Zvi commented:
Changes following the review: Added an enum *setupWork* to specify the type of work (Build, Probe or check-nulls). 
   Removed an unused parameter `isProbe` in `setupGetHash`.
  Extracted the "next" logic from the `isKeyMatch` method, which simplified few things, like no need for a global "index holder".  This works now similar to an iterator, by calling the new `nextLinkInHashChain` method. (Could not implement an iterator because the "next" jumps from one BatchHolder to another).
   And added some more comments.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/common/HashTableTemplate.java:
```diff
@@ -661,16 +669,15 @@ public PutStatus put(int incomingRowIdx, IndexPointer htIdxHolder, int hashCode,
-
+    for ( int currentIndex = startIdx;
+          currentIndex != EMPTY_SLOT;
+          currentIndex = lastEntryBatch.nextLinkInHashChain(currentIndex)) {
```
Presumably this function call gets inlined by the JVM ?  Since it gets called for every Put operation, this is a performance critical code.  Also, the comment in the method says 'this method should only be called following a False from isKeyMatch()'.  It would have been better to call this in the 'else' portion after the 'if' check on line 679 even though the loop condition is being checked after a False is returned (but it is not intuitive).

Ben-Zvi replied:
Given that this method is so simple and short, and being called a lot - it is the ultimate candidate for JVM inlining.
I initially tried to build an iterator for this use, but it gets very complicated as the next links jump from one BatchHolder to another; still the for() loop tries to show a classic traversal of a linked chain; moving the "next" method inside the body spoils this "iterator like" logic.
    The comment should have also said "never call this when currentIndex == EMPTY_SLOT", which is trivial when used in the for().
    I reworded that comment to reflect the "iterator like" use.

Ben-Zvi commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/common/ChainedHashTable.java:
```diff
@@ -121,6 +130,8 @@
   private RecordBatch incomingProbe;
   private final RecordBatch outgoing;
 
+  private enum setupWork {DO_BUILD, DO_PROBE, CHECK_BOTH_NULLS};
```
Done.

Ben-Zvi commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/common/ChainedHashTable.java:
```diff
@@ -253,28 +277,29 @@ private void setupIsKeyMatchInternal(ClassGenerator<HashTable> cg, MappingSet in
 
     // All key expressions compared equal, so return TRUE
-    cg.getEvalBlock()._return(JExpr.TRUE);
+    cg.getEvalBlock()._return(finalResult);
```
Very observant !  The comment was re-written to be more general.

Ben-Zvi commented:
Comments were addressed; no actual code change.

arina-ielchiieva commented:
@Ben-Zvi any particular reason why you have merged three commits instead of doing the squash into one?
